### PR TITLE
82970 Support for multiple workers

### PR DESCRIPTION
**Pull request by djschlegel2**
Basic working version of multi-worker config support plus some cleanups 

1. basho_bench_worker_sup
* now looks for WorkerTypes (named configs for different drivers with local configs) and Workers (specifies numbers of workers of each type (equivalent to concurrent). 
* generates worker start_link commands with triplet ID (WorkerType, WorkerId (within worker type) and GlobalId (unique number for every worker) and copy of worker-type config as LocalConfig. (Just noticed how ugly "GlobalId" is)
* Continues to support original non-worker style configs but marks these with WorkerType = no_workers

2. basho_bench_worker
* handles new start_link/3 format supporting triplet ID and localconfig
* init handles new triplet ID and local config, saves config into process dictionary through basho_bench_config
* includes OptionsConfigs (STILL THINKING ABOUT)
* changes in worker_next_op2 calls

3. basho_bench_stats
* changes in getting name/op information for statistics that recognizes worker type extended names
* cleanups

4. basho_bench_config
* support storing local_config into process dictionary using set_local_config and searching using modified versions of get
* cleanups, renamed state

### CODE REVIEW

File: src/basho_bench_config.erl [hunk 1]
```diff
@@ -28,21 +28,38 @@
...
-         normalize_ips/2,
-         set/2,
-         get/1, get/2]).
+-export([
+    load/1,
+    normalize_ips/2,
+    set/2,
+    get/1,
+    get/2,
+    set_local_config/1
```
Comment (ksnavely) [inline src/basho_bench_config.erl]: Since we're changing the exports, let's alphabetize them.

File: src/basho_bench_config.erl [hunk 2]
```diff
@@ -28,21 +28,38 @@
...
+    terminate/2,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
 
 -include("basho_bench.hrl").
 
--record(basho_bench_config_state, {}).
```
Comment (ksnavely) [inline src/basho_bench_config.erl:43]: Is the atom name change necessary?

File: src/basho_bench_config.erl [hunk 3]
```diff
@@ -51,32 +68,69 @@
...
 start_link() ->
     gen_server:start_link({global, ?MODULE}, ?MODULE, [], []).
 
 
 load(Files) ->
     ensure_started(),
     gen_server:call({global, ?MODULE}, {load_files, Files}). 
-    
+
+
```
Comment (ksnavely) [inline src/basho_bench_config.erl:80]: We should avoid unnecessary whitespace/style changes and try to follow the style the basho bench has established.

File: src/basho_bench_config.erl [hunk 4]
```diff
@@ -51,32 +68,69 @@
...
+set_local_config(LocalConfig) when is_list(LocalConfig) ->
+    Map = my_map_from_list(LocalConfig),
+    set_local_config(Map);
+set_local_config(LocalConfig) when is_map(LocalConfig) ->
+    erlang:put(local_config, LocalConfig).
+
+%% TODO: Change to maps:from_list(List) in Erlang 18
+my_map_from_list(List) ->
```
Comment (ksnavely) [inline src/basho_bench_config.erl]: you should just call this `map_from_list`

File: src/basho_bench_config.erl [hunk 5]
```diff
@@ -89,19 +143,35 @@ get(Key, Default) ->
 %%     {"127.0.0.1", 8092},
 %%     {"127.0.0.1", 8093}]
 normalize_ips(IPs, DefultPort) ->
-    F = fun(Entry, Acc) ->
+    lists:foldl(
+        fun(Entry, Acc) ->
                 normalize_ip_entry(Entry, Acc, DefultPort)
-        end,
-    lists:foldl(F, [], IPs).
```
Comment (ksnavely) [inline src/basho_bench_config.erl:95]: This change doesn't seem necessary.

File: src/basho_bench_stats.erl [hunk 6]
```diff
@@ -35,13 +37,15 @@
...
-                 last_warn = {0,0,0}}).
+-record(state, {
+    ops,
+    start_time = os:timestamp(),
+    last_write_time = os:timestamp(),
+    report_interval,
+    errors_since_last_report = false,
+    stats_writer, stats_writer_data,
+    last_warn = {0,0,0}
+}).
```
Comment (ksnavely) [inline src/basho_bench_stats.erl]: Unnecessary change.

File: src/basho_bench_stats.erl [hunk 7]
```diff
@@ -333,7 +333,64 @@ consume_report_msgs() ->
...
     end.
 
+
 % Assuming all stats sink modules are prefixed with basho_bench_stats_writer_
 normalize_name(StatsSink) when is_atom(StatsSink) ->
     {ok, list_to_atom("basho_bench_stats_writer_" ++ atom_to_list(StatsSink))};
 normalize_name(StatsSink) -> {error, {StatsSink, invalid_name}}.
+
+
+% Build up the list of operations taking into account the user of worker groups
```
Comment (ksnavely) [inline src/basho_bench_stats.erl]: 'use' of worker groups

File: src/basho_bench_stats.erl [hunk 8]
```diff
@@ -333,7 +333,64 @@ consume_report_msgs() ->
...
+% If workers are not used, use the global operations list
+% If workers are being used, use worker_type as an operation prefix,
+%    using either global or per-worker operations depending on the situation.
+%
+% TODO: Think about checking for 0-counts to make it easier to reconfig runs
+% TODO: Understand how undocumented Labels are being used
+
+get_active_ops() ->
```
Comment (ksnavely) [inline src/basho_bench_stats.erl]: active? This seems like a good thing to call `get_worker_ops/0`

File: src/basho_bench_stats.erl [hunk 9]
```diff
@@ -333,7 +333,64 @@ consume_report_msgs() ->
...
+                    ({Label, OpTag, _Count}) -> {Label, OpTag};
+                    ({Label, OpTag, _Count, _OptionsList}) -> {Label, OpTag}
+                end, basho_bench_config:get(operations, []));
+       %% Use workers to determine active worker types for current config
+        _ -> 
+            get_worker_ops(Workers, basho_bench_config:get(worker_types), [])
+    end.
+
+
+get_worker_ops([],_WorkerTypes, ACC) ->
```
Comment (ksnavely) [inline src/basho_bench_stats.erl]: `Acc` is the typical style.

Comment (ksnavely) [reply src/basho_bench_stats.erl]: Watch your style, missing space here.

File: src/basho_bench_stats.erl [hunk 10]
```diff
@@ -333,7 +333,64 @@ consume_report_msgs() ->
...
+     ACC;
+get_worker_ops(Workers, WorkerTypes, ACC) ->
+     [ WorkerEntry | RestWorkers ] = Workers,
```
Comment (ksnavely) [inline src/basho_bench_stats.erl]: You have errant spaces before/after brackets etc. here and below.

File: src/basho_bench_stats.erl [hunk 11]
```diff
@@ -333,7 +333,64 @@ consume_report_msgs() ->
...
+                { worker_op_name(WorkerType,OpTag), worker_op_name(WorkerType,OpTag)};
+           ({Label, OpTag, _Count2}) ->
+                { worker_op_name(WorkerType,Label), worker_op_name(WorkerType,OpTag)};
+           ({Label, OpTag, _Count2, _OptionsList}) ->
+                { worker_op_name(WorkerType,Label), worker_op_name(WorkerType,OpTag)}
+        end, WorkerOps),
+     get_worker_ops(RestWorkers, WorkerTypes, ACC++Ops).
+
+
+%% QUESTION: Generating atoms is not ideal - better way to handle this ? Needed for stats output too
```
Comment (ksnavely) [inline src/basho_bench_stats.erl]: Seems OK for this I think...

File: src/basho_bench_worker.erl [hunk 12]
```diff
@@ -56,20 +60,20 @@
 %% API
 %% ====================================================================
 
-start_link(SupChild, Id) ->
-    case basho_bench_config:get(distribute_work, false) of
```
Comment (ksnavely) [inline src/basho_bench_worker.erl:60]: Again, I think we have to be extremely careful about style changes, even though it's _normally_ the right thing to do. @chewbranca opinion?

File: src/basho_bench_worker.erl [hunk 13]
```diff
@@ -99,20 +111,31 @@ init([SupChild, Id]) ->
             now -> now()
         end,
 
-    RngSeed = {A1+Id, A2+Id, A3+Id},
+    %% TODO: Do we want WorkerId (i.e. per worker type) or WorkerGlobalId (unique) ?
```
Comment (ksnavely) [inline src/basho_bench_worker.erl]: Unique would be the best I believe.

File: src/basho_bench_worker.erl [hunk 14]
```diff
@@ -180,30 +202,58 @@ terminate(_Reason, _State) ->
...
+
+ops_configs(Id, Operations) ->
+    ops_configs(Id, Operations, []).
+
+%% Process individual operation entries in Operations list
+ops_configs(_Id, [], ACC) ->
+    ACC;
+ops_configs(Id, [{_Label, OpTag, _Count, OptionsList}|RestOps], ACC) ->
+    ops_configs(Id, RestOps, [ {OpTag, ops_configs_list(Id, OptionsList, [])} | ACC]);
+ops_configs(Id, [_Op|RestOps], ACC) ->
```
Comment (ksnavely) [inline src/basho_bench_worker.erl]: So do we expect operations to not match the previous function head?

File: src/basho_bench_worker.erl [hunk 15]
```diff
@@ -180,30 +202,58 @@ terminate(_Reason, _State) ->
...
+    ops_configs(Id, RestOps, ACC).
+
+% Process individual config parameter entries within an Operations OptionsList
```
Comment (ksnavely) [inline src/basho_bench_worker.erl]: 'Process' is pretty generic. It looks like this function generates a keygen or valgen for each worker? Why is this needed? We still call `init_generators` yeah?

File: src/basho_bench_worker.erl [hunk 16]
```diff
@@ -246,30 +296,38 @@ worker_idle_loop(State) ->
...
+   OpConfig = proplists:get_value(OpTag, State#state.ops_configs),
+   KeyGen = proplists:get_value(keygen, State#state.keygen, OpConfig),
+   ValueGen = proplists:get_value(valuegen, State#state.valgen, OpConfig),
+   catch (State#state.driver):run(OpTag, KeyGen, ValueGen,
+                                  State#state.driver_state).
 
 worker_next_op(State) ->
     Next = element(random:uniform(State#state.ops_len), State#state.ops),
-    {_Label, OpTag} = Next,
+    {Label, OpTag} = Next,
```
Comment (ksnavely) [inline src/basho_bench_worker.erl]: Skip the intermediate variable `{Label, OpTag} = element(random:uniform(State#state.ops_len), State#state.ops)`

File: src/basho_bench_worker_sup.erl [hunk 17]
```diff
@@ -71,6 +92,23 @@ worker_specs(0, Acc) ->
     Acc;
 worker_specs(Count, Acc) ->
     Id = list_to_atom(lists:concat(['basho_bench_worker_', Count])),
-    Spec = {Id, {basho_bench_worker, start_link, [Id, Count]},
-            transient, 5000, worker, [basho_bench_worker]},
+    %% Use "no_workers" atom for original non-worker case
+    Spec = {Id, {basho_bench_worker, start_link, [Id, {no_workers, Count, Count}, []]},
```
Comment (ksnavely) [inline src/basho_bench_worker_sup.erl]: It's not really no workers though, right, it's just one worker type?

Comment (ksnavely) [reply src/basho_bench_worker_sup.erl]: I see, this refers to the `workers` config item? `no_workers` is still a bit confusing.

File: src/basho_bench_config.erl (hunk 2 above)
Comment (djschlegel2) [reply src/basho_bench_config.erl:43]: That was part of @chewbranca's cleanup changes before he waded into rearranging my prototype code.

File: src/basho_bench_config.erl (hunk 3 above)
Comment (djschlegel2) [reply src/basho_bench_config.erl:80]: In that case it IS following the whitespace style of the neighboring functions, was just inconsistent

File: src/basho_bench_worker_sup.erl (hunk 17 above)
Comment (djschlegel2) [reply src/basho_bench_worker_sup.erl]: Would 'single_worker' make any more sense ?

File: src/basho_bench_config.erl (hunk 5 above)
Comment (djschlegel2) [reply src/basho_bench_config.erl:95]: part of @chewbranca's cleanup changes

File: src/basho_bench_worker.erl (hunk 12 above)
Comment (djschlegel2) [reply src/basho_bench_worker.erl:60]: Not sure which style change you are referring to - this was a functional change to accept a third parameter, WorkerConf

File: src/basho_bench_config.erl [hunk 18]
```diff
@@ -51,33 +68,64 @@
...
+
+set_local_config(LocalConfig) when is_list(LocalConfig) ->
+    Map = map_from_list(LocalConfig),
+    set_local_config(Map);
+set_local_config(LocalConfig) when is_map(LocalConfig) ->
+    erlang:put(local_config, LocalConfig).
+
+%% TODO: Change to maps:from_list(List) in Erlang 18
+map_from_list(List) ->
+    map_from_list(List, #{}).
```
Comment (chewbranca) [inline src/basho_bench_config.erl]: There should be two newlines around this function on each side.

File: src/basho_bench_config.erl [hunk 19]
```diff
@@ -117,44 +181,51 @@ normalize_ip_entry(IP, Normalized, DefaultPort) ->
...
+    KVs = [
     case file:consult(File) of
         {ok, Terms} ->
             Terms;
         {error, Reason} ->
             ?FAIL_MSG("Failed to parse config file ~s: ~p\n", [File, Reason]),
             throw(invalid_config),
             notokay
-    end || File <- Files ],
+    end || File <- Files],
```
Comment (chewbranca) [inline src/basho_bench_config.erl]: List comprehensions should not be more than one line, use lists:map instead.

File: src/basho_bench_stats.erl [hunk 20]
```diff
@@ -333,7 +332,59 @@ consume_report_msgs() ->
...
+     %% Get either per-worker ops or global ops
+     WorkerOps = proplists:get_value(operations, WorkerConfig),
+
+     Ops = lists:map(
+        fun({OpTag, _Count2}) -> 
+                { worker_op_name(WorkerType, OpTag), worker_op_name(WorkerType, OpTag)};
+           ({Label, OpTag, _Count2}) ->
+                { worker_op_name(WorkerType, Label), worker_op_name(WorkerType, OpTag)};
+           ({Label, OpTag, _Count2, _OptionsList}) ->
+                { worker_op_name(WorkerType, Label), worker_op_name(WorkerType, OpTag)}
```
Comment (chewbranca) [inline src/basho_bench_stats.erl]: What was the verdict here? Remove the more complicated ops config support?

File: src/basho_bench_worker.erl [hunk 21]
```diff
@@ -99,23 +105,27 @@ init([SupChild, Id]) ->
...
         end,
 
-    RngSeed = {A1+Id, A2+Id, A3+Id},
+    RngSeed = {A1 + WorkerGlobalId, A2 + WorkerGlobalId, A3 + WorkerGlobalId},
 
     %% Pull all config settings from environment
     Driver  = basho_bench_config:get(driver),
-    Ops     = ops_tuple(),
+    Operations = basho_bench_config:get(operations),
+    Ops     = ops_tuple(Operations),
```
Comment (chewbranca) [inline src/basho_bench_worker.erl]: Extra space between `Ops` and `=`.

File: src/basho_bench_worker_sup.erl [hunk 22]
```diff
@@ -60,8 +60,28 @@ init([]) ->
     %% specs for each
 
     basho_bench_profiler:maybe_start_profiler(basho_bench_config:get(enable_profiler, false)),
-    Workers = worker_specs(basho_bench_config:get(concurrent), []),
-    {ok, {{one_for_one, 5, 10}, Workers}}.
+
+    Workers = basho_bench_config:get(workers, []),
+    WorkerTypes = basho_bench_config:get(worker_types, []),
```
Comment (chewbranca) [inline src/basho_bench_worker_sup.erl]: We should use the one arity config get here so that it explodes in the even workers is set and worker types is not set.

File: src/basho_bench_worker_sup.erl [hunk 23]
```diff
@@ -71,6 +91,23 @@ worker_specs(0, Acc) ->
...
+            Id = list_to_atom(lists:concat(
+                ['basho_bench_worker_', WorkerType, '_', I])),
+            Spec = {
+                Id,
+                {basho_bench_worker, start_link, [Id, {WorkerType, I, I+BaseGlobalId}, Conf]},
+                transient, 5000, worker, [basho_bench_worker]},
+            AccP ++ [Spec]
+        end,
+        [], lists:seq(1, Count)),
+    worker_specs_multi(Rest, BaseGlobalId+Count, Acc0 ++ Acc).
```
Comment (chewbranca) [inline src/basho_bench_worker_sup.erl]: The separation between init and these functions is a bit awkward. I would switch it up to be something like:


```erlang
init([]) ->
    %% Get the number concurrent workers we're expecting and generate child
    %% specs for each

    basho_bench_profiler:maybe_start_profiler(basho_bench_config:get(enable_profiler, false)),

    WorkerSpecs = worker_specs(basho_bench_config:get(workers, [])),
    {ok, {{one_for_one, 5, 10}, WorkerSpecs}}.


%% ===================================================================
%% Internal functions
%% ===================================================================


worker_specs([]) ->
    worker_specs_single(basho_bench_config:get(concurrent), []);
worker_specs(Workers) ->
    WorkerConfs = combine_mutli_confs(Workers, basho_bench_config:get(worker_types)),
    worker_specs_multi(WorkerConfs, 0, []).


worker_specs_single(0, Acc) ->
    Acc;
worker_specs_single(Count, Acc) ->
    Id = list_to_atom(lists:concat(['basho_bench_worker_', Count])),
    %% Use "single_worker" atom for original non-worker case
    Spec = {Id, {basho_bench_worker, start_link, [Id, {single_worker, Count, Count}, []]},
                 transient, 5000, worker, [basho_bench_worker]},
    worker_specs_single(Count-1, [Spec | Acc]).


worker_specs_multi([], _BaseGlobalId, Acc) ->
    Acc;
worker_specs_multi([{WorkerType, Count, Conf} | Rest], BaseGlobalId, Acc0) ->
    Acc = lists:foldl(
        fun(I, AccP) ->
            Id = list_to_atom(lists:concat(
                ['basho_bench_worker_', WorkerType, '_', I])),
            Spec = {
                Id,
                {basho_bench_worker, start_link, [Id, {WorkerType, I, I+BaseGlobalId}, Conf]},
                transient, 5000, worker, [basho_bench_worker]},
            [Spec | AccP]
        end,
        Acc0, lists:seq(1, Count)),
    worker_specs_multi(Rest, BaseGlobalId+Count, Acc).


combine_multi_confs(Workers, WorkerTypes) ->
            lists:map(
                fun({WT, Count}) ->
                    %% Burn in {concurrent, Count} to the WorkerConf
                    %% Key-generator(sequential) needs to know number of workers sharing that keygen
                    Conf0 = proplists:get_value(WT, WorkerTypes, []),
                    Conf = [{concurrent, Count} | proplists:delete(concurrent, Conf0)]),
                    {WT, Count, Conf}
                end, Workers).

```

File: src/basho_bench_worker_sup.erl [hunk 24]
```diff
@@ -71,6 +91,23 @@ worker_specs(0, Acc) ->
...
+worker_specs_multi([{WorkerType, Count, Conf} | Rest], BaseGlobalId, Acc0) ->
+    Acc = lists:foldl(
+        fun(I, AccP) ->
+            Id = list_to_atom(lists:concat(
+                ['basho_bench_worker_', WorkerType, '_', I])),
+            Spec = {
+                Id,
+                {basho_bench_worker, start_link, [Id, {WorkerType, I, I+BaseGlobalId}, Conf]},
+                transient, 5000, worker, [basho_bench_worker]},
+            AccP ++ [Spec]
```
Comment (chewbranca) [inline src/basho_bench_worker_sup.erl]: Concatenating lists like this is usually not what you want, as it's quite inefficient. You can flip this around and do `[Spec | AccP]` which is quite efficient. The only time you should use `++` is for combining strings (and even then don't do it in a tight loop).

**Comment by chewbranca** [comment]
@ksnavely RE whitespace and cleanup changes, this PR is cleaning up a halfway completed PR from last year, so a bit of the style comments were to address things from then. I had originally made the style changes as a dedicated commit, which isn't as clean as a dedicated PR, but at this point I think the cleanup changes help here and I think it makes sense to land them with this PR as a special consideration.

File: src/basho_bench_stats.erl (hunk 20 above)
Comment (djschlegel2) [reply src/basho_bench_stats.erl]: That was what I did, I yanked it. It was written to try to support per-operation parameters but that would run against the basic mechanism of basho_bench_config:get retrieving values from the process_dict or global environment to work. 

If it is mainly keygen and valgen that we need, we now have both per-worker config and api-state based named keygens to provide a lot more functionality than we had. I filed FB 84560 with the code that I yanked out so we didn't loose it but it wasn't working as implemented. If we need it it would be helpful to have some examples to clarify how it needs to work.

Comment (djschlegel2) [reply src/basho_bench_stats.erl]: And I will remove this now useless clause

File: src/basho_bench_worker_sup.erl (hunk 22 above)
Comment (djschlegel2) [reply src/basho_bench_worker_sup.erl]: No. Right idea but the wrong spot. We want to tolerate both Workers and WorkerTypes being undefined to support the traditional case. But if Workers is defined and there isn't a matching WorkerTypes entry we want THAT to blow up, below in line 77 so I am changing that to be where it fails. That would also cover WorkerTypes being undefined. 

                                      Conf = proplists:get_value(WT, WorkerTypes),

File: src/basho_bench_worker.erl [hunk 25]
```diff
@@ -243,33 +253,38 @@ worker_idle_loop(State) ->
...
 worker_next_op(State) ->
-    Next = element(random:uniform(State#state.ops_len), State#state.ops),
-    {_Label, OpTag} = Next,
+    {Label, OpTag} = element(random:uniform(State#state.ops_len), State#state.ops),
     Start = os:timestamp(),
     Result = worker_next_op2(State, OpTag),
     ElapsedUs = erlang:max(0, timer:now_diff(os:timestamp(), Start)),
+
+    OpName = { basho_bench_stats:worker_op_name(State#state.worker_type, Label),
+               basho_bench_stats:worker_op_name(State#state.worker_type, OpTag)},
```
Comment (chewbranca) [inline src/basho_bench_worker.erl:271]: Minor nit, incorrect indentation here. This should be:

```erlang
OpName = {
    basho_bench_stats:worker_op_name(State#state.worker_type, Label),
    basho_bench_stats:worker_op_name(State#state.worker_type, OpTag)
},
```

**Comment by chewbranca** [comment]
+1
